src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/MessageInterception/DefaultInterceptorManager.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+#nullable enable
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage.MessageInterception
+{
+    [Export(typeof(InterceptorManager))]
+    internal sealed class DefaultInterceptorManager : InterceptorManager
+    {
+        private readonly IList<Lazy<MessageInterceptor, IInterceptMethodMetadata>> _lazyInterceptors;
+
+        [ImportingConstructor]
+        public DefaultInterceptorManager([ImportMany] IEnumerable<Lazy<MessageInterceptor, IInterceptMethodMetadata>> lazyInterceptors)
+        {
+            _ = lazyInterceptors ?? throw new ArgumentNullException(nameof(lazyInterceptors));
+            _lazyInterceptors = lazyInterceptors.ToList();
+        }
+
+        public override bool HasInterceptor(string messageName)
+        {
+            foreach (var interceptor in _lazyInterceptors)
+            {
+                foreach (var method in interceptor.Metadata.InterceptMethods)
+                {
+                    if (method.Equals(messageName, StringComparison.Ordinal))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public override async Task<JToken?> ProcessInterceptorsAsync(string methodName, JToken token, string sourceLanguageName, CancellationToken cancellationToken)
+        {
+            for (var i = 0; i < _lazyInterceptors.Count; i++)
+            {
+                var item = _lazyInterceptors[i];
+                if (CanInterceptMessage(methodName, item.Metadata.InterceptMethods))
+                {
+                    var result = await item.Value.ApplyChangesAsync(token, sourceLanguageName, cancellationToken);
+                    if (result.UpdatedToken is null)
+                    {
+                        // The interceptor has blocked this message
+                        return null;
+                    }
+
+                    token = result.UpdatedToken;
+
+                    if (result.ChangedDocumentUri)
+                    {
+                        // If the DocumentUri changes, we need to restart the loop
```

Comment:
Technically this can cause us to infinite loop here. Don't necessarily need to solve this now but food for thought when this becomes more mainstream.

Reply:
True, and @ToddGrun pointed this out in the original PR.  That would require a bad/buggy actor or interaction between two actors, to cause the document URI to be constantly changed.  It's definitely worth keeping in mind though as you both pointed out.

I though about using (or deriving from) [IOrderable](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.utilities.iorderable?view=visualstudiosdk-2019) but that then required that either the contained or host language know about every party on the other side (e.g. either CSS needs to know about HTML, Razor, Django, Python, etc.., or all of those need to know about CSS; repeat for each content type, and in this scenario HTML needs to always know both sides).  That turned messy and hard-to-maintain rather quickly.